shared/subspace-object-fetching/src/object_fetcher.rs

```diff
@@ -0,0 +1,402 @@
+        next_piece_index += PieceIndex::ONE;
+        read_records_data.extend_from_slice(&raw_piece_bytes);
+
+        // Let's see how many bytes encode compact length encoding of the data, see
+        // https://docs.substrate.io/reference/scale-codec/#fn-1 for details.
+        let data_length_bytes_length: u32 = match read_records_data[mapping.offset as usize] % 4 {
+            0 => 1,
+            1 => 2,
+            2 => 4,
+            _ => {
+                // Big integer mode, for values 2^30 and higher, but we only support MAX_OBJECT_SIZE
+                let error_string = format!(
+                    "Invalid data length prefix found: 0x{:02x} for object {mapping:?}",
+                    read_records_data[mapping.offset as usize]
+                );
+                error!(error = %error_string);
+
+                return Err(Error::StringError(error_string));
+            }
+        };
+
+        let offset_after_object = (mapping.offset + data_length_bytes_length) as usize;
+        // Same as `before_last_two_bytes`, but accounts for compact encoding of data length
+        let length_before_last_two_bytes = offset_after_object < RawRecord::SIZE - 1 - 2;
+        // Similar to `length_before_last_two_bytes`, but uses the whole record if needed
+        let length_before_record_end = offset_after_object < RawRecord::SIZE - 1;
+
+        let data_length_result = if length_before_last_two_bytes {
+            Compact::<u32>::decode(&mut &read_records_data[mapping.offset as usize..])
+        } else if !last_data_piece_in_segment {
+            if !length_before_record_end {
+                // Need the next piece to read the length of data
+                let raw_piece_bytes = self.read_and_decode_piece(next_piece_index).await?;
+                next_piece_index += PieceIndex::ONE;
+                read_records_data.extend_from_slice(&raw_piece_bytes);
+            }
+
+            Compact::<u32>::decode(&mut &read_records_data[mapping.offset as usize..])
+        } else {
+            // Super fast read is not possible
+            return Ok(None);
+        };
+
+        let Compact(data_length) = data_length_result.map_err(|error| {
+            let error_string =
+                format!("Failed to read object data length: {error}, for object {mapping:?}");
+            error!(error = %error_string);
+
+            Error::StringError(error_string)
+        })?;
+
+        if u64::from(data_length_bytes_length + data_length) > bytes_available_in_segment {
+            // Not enough data without crossing segment boundary
+            return Ok(None);
+        }
+
+        let mut data = read_records_data
+            [mapping.offset as usize + data_length_bytes_length as usize..]
+            .to_vec();
+        drop(read_records_data);
+
+        // Read more pieces until we have enough data
+        while data.len() <= data_length as usize {
```

Comment:
What kind of optimisation?

If you’re asking to parallelise piece retrieval, we can use the same “get a list of pieces” method for both this case and the segment case.

Reply:
This can retrieve up to 5 * 1 MB pieces, providing 5 * 992 kB = ~4.9 MB of raw data.

But it seems like we might only be getting 1-4 pieces here, because of other limits on the size of `system.remark`.

Reply:
Makes sense, I've added a TODO comment.